framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -1641,87 +1641,93 @@ && getCurrentPath().getParentPath().getLeaf().getKind()
+                    }
+                    break;
+                case EXCEPTION_PARAMETER:
+                    if (locations.contains(TypeUseLocation.EXCEPTION_PARAMETER)) {
+                        issueError = false;
+                    }
+                    break;
+                case PARAMETER:
+                    if (tree.getName().contentEquals("this")) {
+                        if (locations.contains(TypeUseLocation.RECEIVER)) {
                             issueError = false;
                         }
-                        break;
-                    case ENUM_CONSTANT:
-                        if (locations.contains(TypeUseLocation.FIELD)
-                                || locations.contains(TypeUseLocation.CONSTRUCTOR_RESULT)) {
+                    } else {
+                        if (locations.contains(TypeUseLocation.PARAMETER)) {
                             issueError = false;
                         }
-                        break;
-                    default:
-                        throw new BugInCF("Location not matched");
-                }
-                if (issueError) {
-                    checker.reportError(
-                            tree,
-                            "type.invalid.annotations.on.location",
-                            am.toString(),
-                            element.getKind().name());
-                }
+                    }
+                    break;
+                case RESOURCE_VARIABLE:
+                    if (locations.contains(TypeUseLocation.RESOURCE_VARIABLE)) {
+                        issueError = false;
+                    }
+                    break;
+                case FIELD:
+                    if (locations.contains(TypeUseLocation.FIELD)) {
+                        issueError = false;
+                    }
+                    break;
+                case ENUM_CONSTANT:
+                    if (locations.contains(TypeUseLocation.FIELD)
+                            || locations.contains(TypeUseLocation.CONSTRUCTOR_RESULT)) {
+                        issueError = false;
+                    }
+                    break;
+                default:
+                    throw new BugInCF("Location not matched");
+            }
+            if (issueError) {
+                checker.reportError(
+                        tree,
+                        "type.invalid.annotations.on.location",
+                        am.toString(),
+                        element.getKind().name());
             }
         }
     }
 
     /**
-     * Validate if the annotations on the tree are at the right locations, which are specified by
-     * the meta-annotation @TargetLocations.
+     * Validate if the qualifiers on the tree are at the right locations, which are specified by the
+     * meta-annotation @TargetLocations.
      *
-     * @param tree annotations on this VariableTree will be validated
+     * @param tree qualifiers on this VariableTree will be validated
```

Comment:
`tree` is a `Tree`. Should it be changed to `VariableTree`?

Reply:
thanks for pointing this out! The comment is misleading as it could accept a `Tree` parameter. I have updated this comment.